src/format/mod.rs

```diff
@@ -720,52 +709,56 @@ impl Piece {
 }
 
 /// Wrapper struct for formatting time with the provided format string.
-pub(crate) struct TimeFormatter<'t, 'f, T> {
+pub(crate) struct TimeFormatter<'t, 'f, T, F: ?Sized> {
     /// Time implementation
     time: &'t T,
     /// Format string
-    format: &'f [u8],
+    format: &'f F,
 }
 
-impl<'t, 'f, T: CheckedTime> TimeFormatter<'t, 'f, T> {
+impl<'t, 'f, T, F> TimeFormatter<'t, 'f, T, F>
+where
+    T: CheckedTime,
+    F: AsRef<[u8]> + ?Sized + Index<RangeTo<usize>, Output = F>,
+    GenericCursor<'f, F>: Cursor<'f, Slice = F>,
+{
     /// Construct a new `TimeFormatter` wrapper.
-    pub(crate) fn new<F: AsRef<[u8]> + ?Sized>(time: &'t T, format: &'f F) -> Self {
-        Self {
-            time,
-            format: format.as_ref(),
-        }
+    pub(crate) fn new(time: &'t T, format: &'f F) -> Self {
+        Self { time, format }
     }
 
     /// Format time using the format string.
-    pub(crate) fn fmt(&self, buf: &mut dyn Write) -> Result<(), Error> {
+    pub(crate) fn fmt(&self, buf: &mut impl Write<Slice = F>) -> Result<(), Error> {
         // Do nothing if the format string is empty
-        if self.format.is_empty() {
+        if self.format.as_ref().is_empty() {
             return Ok(());
         }
 
         // Use a size limiter to limit the maximum size of the resulting
         // formatted string
-        let size_limit = self.format.len().saturating_mul(512 * 1024);
+        let size_limit = self.format.as_ref().len().saturating_mul(512 * 1024);
         let mut f = SizeLimiter::new(buf, size_limit);
 
-        let mut cursor = Cursor::new(self.format);
+        let mut cursor = GenericCursor::new(self.format);
 
         loop {
-            f.write_all(cursor.read_until(|&x| x == b'%'))?;
+            f.write_all(cursor.read_until(|x| x == b'%'))?;
 
-            let remaining_before = cursor.remaining();
-
-            // Read the '%' character
-            if cursor.next().is_none() {
+            if cursor.remaining().as_ref().is_empty() {
                 break;
             }
 
+            let remaining_before = cursor.remaining();
+
+            // Read the `%` character
+            cursor.advance();
+
             match Self::parse_spec(&mut cursor)? {
                 Some(piece) => piece.fmt(&mut f, self.time)?,
                 None => {
                     // No valid format specifier was found
-                    let remaining_after = cursor.remaining();
-                    let text = &remaining_before[..remaining_before.len() - remaining_after.len()];
+                    let remaining_len = cursor.remaining().as_ref().len();
+                    let text = &remaining_before[..remaining_before.as_ref().len() - remaining_len];
```

Comment:
Just talking through this to verify my understanding.

This line is what the generic bound on `RangeTo` is for.

And for `str`, because we're always reading until a `char` boundary, this indexing is guaranteed not to panic.

Reply:
Yes this is correct.